folding-schemes/src/folding/nova/ova.rs

```diff
@@ -0,0 +1,271 @@
+        &self,
+        params: &CS::ProverParams,
+        x: Vec<C::ScalarField>,
+        t_or_e: Vec<C::ScalarField>,
+    ) -> Result<CommittedInstance<C>, Error> {
+        let cmWE = CS::commit(params, &[self.w.clone(), t_or_e].concat(), &self.rW)?;
+        Ok(CommittedInstance {
+            u: C::ScalarField::one(),
+            cmWE,
+            x,
+        })
+    }
+}
+
+impl<C: CurveGroup> Dummy<&R1CS<CF1<C>>> for Witness<C> {
+    fn dummy(r1cs: &R1CS<CF1<C>>) -> Self {
+        Self {
+            w: vec![C::ScalarField::zero(); r1cs.A.n_cols - 1 - r1cs.l],
+            rW: C::ScalarField::zero(),
+        }
+    }
+}
+
+/// Implements the NIFS (Non-Interactive Folding Scheme) trait for Ova.
+pub struct NIFS<C: CurveGroup, CS: CommitmentScheme<C, H>, const H: bool = false> {
+    _c: PhantomData<C>,
+    _cp: PhantomData<CS>,
+}
+
+impl<C: CurveGroup, CS: CommitmentScheme<C, H>, const H: bool> NIFSTrait<C, CS, H>
+    for NIFS<C, CS, H>
+where
+    <C as Group>::ScalarField: Absorb,
+    <C as CurveGroup>::BaseField: PrimeField,
+{
+    type CommittedInstance = CommittedInstance<C>;
+    type Witness = Witness<C>;
+    type ProverAux = ();
+    type VerifierAux = ();
+
+    fn new_witness(w: Vec<C::ScalarField>, _e_len: usize, rng: impl RngCore) -> Self::Witness {
+        Witness::new::<H>(w, rng)
+    }
+
+    fn new_instance(
+        W: &Self::Witness,
+        params: &CS::ProverParams,
+        x: Vec<C::ScalarField>,
+        aux: Vec<C::ScalarField>, // t_or_e
+    ) -> Result<Self::CommittedInstance, Error> {
+        W.commit::<CS, H>(params, x, aux)
+    }
+
+    fn fold_witness(
+        r: C::ScalarField, // in Ova's hackmd denoted as `alpha`
+        W_i: &Self::Witness,
+        w_i: &Self::Witness,
+        _aux: &Self::ProverAux,
+    ) -> Result<Self::Witness, Error> {
+        let w: Vec<C::ScalarField> = W_i
+            .w
+            .iter()
+            .zip(&w_i.w)
+            .map(|(a, b)| *a + (r * b))
+            .collect();
+
+        let rW = W_i.rW + r * w_i.rW;
+        Ok(Self::Witness { w, rW })
+    }
+
+    fn compute_aux(
+        _cs_prover_params: &CS::ProverParams,
+        _r1cs: &R1CS<C::ScalarField>,
+        _W_i: &Self::Witness,
+        _U_i: &Self::CommittedInstance,
+        _w_i: &Self::Witness,
+        _u_i: &Self::CommittedInstance,
+    ) -> Result<(Self::ProverAux, Self::VerifierAux), Error> {
+        Ok(((), ()))
+    }
+
+    fn get_challenge<T: Transcript<C::ScalarField>>(
+        transcript: &mut T,
+        pp_hash: C::ScalarField, // public params hash
+        U_i: &Self::CommittedInstance,
+        u_i: &Self::CommittedInstance,
+        _aux: &Self::VerifierAux,
+    ) -> Vec<bool> {
+        // reuse Nova's get_challenge method
+        ChallengeGadget::<C, Self::CommittedInstance>::get_challenge_native(
+            transcript,
+            pp_hash,
+            U_i,
+            u_i,
+            &C::zero(), // empty in Ova's case
```

Comment:
This point later in-circuit is done as a non-native-affine-point, which in case of the point being zero (which is the case in Ova), when converted into field elements (from its coordinates) in order to absorb it in the transcript, and for the 0 case, it takes just 4 constraint-field elements in total. In the context of the AugmentedFCircuit I think that it is an almost-negligible amount of constraints.

Reply:
But if we're worried about those extra constraints (which I think that in the context of the AugmentedFCircuit are quite small),
can put an 'if' in this method, which in the case of Nova it absorbs the cmT and in Ova it does not. So in this way we still avoid duplicating the entire ChallengeGadget and its methods as it was prior to this PR.

Reply:
It's not for the constraints. But that we need to follow the spec as is. I don't want to hash more things as if later someone wants to integrate with this, they will have a hard time finding why they have an error on the transcript verification. Until they realize we are hashing Zero here for no reason.

Reply:
Have you read this one https://github.com/privacy-scaling-explorations/sonobe/pull/165#discussion_r1785813570 ?

Reply:
Done the abstraction that was mentioning at the comment https://github.com/privacy-scaling-explorations/sonobe/pull/165#discussion_r1785813570 in the commit: https://github.com/privacy-scaling-explorations/sonobe/pull/165/commits/31096c8e31c9bad55074f5314b20c227b98633eb
so that we still avoid duplicating the code of the ChallengeGadget struct & methods, while in the Ova case not hashing the zero point.

Reply:
Just saw the code. I like it. Apologies for the confusion.